doc/src/lexer_tutorial/002_raw_delimited_content.md

```diff
@@ -0,0 +1,347 @@
+(`(`, `)`), so it was easy to embed those tokens directly in the grammar,
+as we saw in the earlier sections.
+
+However, some languages do not have such easy separations at the lexical level.
+
+Consider parsing a language with string literals. We will deliberately choose
+one that is very simple: All it can do is bind variables, and the variables are
+always single characters, like this:
+
+```
+x = "a"
+y = "bc"
+```
+
+Using what we have learned so far, we might try a grammar like the below.
+
+```lalrpop
+use super::{Var, Lit, Stmt};
+
+grammar;
+
+pub Var: Var = <r"[x-z]"> => <>.chars().next().unwrap().into();
+
+pub Lit: Lit = "\"" <r"[a-c]*"> "\"" => <>.into();
+
+pub Stmt: Stmt = <Var> "=" <Lit> => (<>).into();
+```
+
+This grammar deliberately uses regular expressions that are very limited in scope,
+ensuring that there is no overlap between variable names and string contents,
+so that we can see what kinds of cases that this approach *does* handle.
+
+We can combine the above with some library code and get a system that seems to work.
+
+```rust
+lalrpop_mod!(pub nobol1); // syntesized by LALRPOP
+
+#[derive(PartialEq, Debug)]
+pub struct Var(char);
+
+#[derive(PartialEq, Debug)]
+pub struct Lit(String);
+
+#[derive(PartialEq, Debug)]
+pub struct Stmt(Var, Lit);
+
+impl From<char> for Var { fn from(c: char) -> Self { Var(c) } }
+impl From<&str> for Lit { fn from(s: &str) -> Self { Lit(s.into()) } }
+impl From<String> for Lit { fn from(s: String) -> Self { Lit(s.into()) } }
+impl From<(Var, Lit)> for Stmt { fn from((v, l): (Var, Lit)) -> Self { Stmt(v, l) } }
+impl From<(char, &str)> for Stmt { fn from((v, l): (char, &str)) -> Self { Stmt(v.into(), l.into()) } }
+
+#[test]
+fn nobol1() {
+    assert_eq!(nobol1::VarParser::new().parse("x"), Ok('x'.into()));
+    assert_eq!(nobol1::LitParser::new().parse(r#""abc""#), Ok("abc".into()));
+    assert_eq!(nobol1::StmtParser::new().parse(r#"x = "a""#), Ok(('x', "a").into()));
+    assert_eq!(nobol1::StmtParser::new().parse(r#"y = "bc""#), Ok(('y', "bc").into()));
+}
+```
+
+However, as soon as we try to handle examples where there *is* overlap between
+the variable names and the string contents, we run into a problem.
+
+Consider an example like this:
+
+```
+z = "xyz"
+```
+
+If we encode that as a test,
+
+```rust
+#[test]
+fn no_ball() {
+    assert_eq!(nobol1::StmtParser::new().parse(r#"z = "xyz""#), Ok(('z', "xyz").into()));
+}
+```
+
+we get the following:
+
+```
+thread 'no_ball' panicked at 'assertion failed: `(left == right)`
+  left: `Err(UnrecognizedToken { token: (5, Token(1, "x"), 6), expected: ["r#\"[a-c]*\"#"] })`,
+ right: `Ok(Stmt(Var('z'), Lit("xyz")))`', doc/nobol/src/main.rs:31:5
+```
+
+Okay, this sounds easy to fix: the regular expression for string literal content should
+be expanded to handle characters other than `a`, `b` and `c`.
+
+So does this work?
+
+```lalrpop
+pub Lit: Lit = "\"" <r"[a-z]*"> "\"" => <>.into();
+```
```

Comment:
I think this section can be cut out. The guide starts by saying that the calculator had a clear separation between lexical tokens, but this document is about grammars with some form of overlapping. However, we start the section with a grammar without overlapping, and which honestly looks really artificial. I'm not sure it adds a lot of value.

My suggestion would be to get rid of this first step and directly starts with imagining a language with strings, variables and statements, and then write the naive ambiguous version of this grammar.

Reply:
Its been over a year, but from what I remember about the way that my brain works, it was easiest for *me* to get the initial grammar working by starting from the artificially limited case, get all those pieces to work, then figure out how to generalize it.

Sure, the way I restricted it was absurdly limited. But there is plenty of precedent, especially in PL work, of people forcing programmers to follow seemingly absurd syntactic restrictions in order to satisfy the limits of the lexer/parser implementation.

Anyway, my point is more that I thought it was worthwhile to spell out "here's working code, but it doesn't readily generalize." And it *is* worthwhile having those artificial limits when it makes the discussion clear; for example, in the section "The right way to do this", the places where it refers to `r"[a-c]*"` is entirely *unambiguous* because that range is solely used in the definition for the `Lit` production.

Reply:
Having said all that, let me take a shot at the re-write you suggest. Maybe it will come out cleaner.

Reply:
Addressed in fb568a3be286438944569aae7d25b33e6e0fa17b